cupy/core/raw.pyx

```diff
@@ -105,6 +105,36 @@ cdef class RawKernel:
             self._kernel_cache[dev] = ker
         return ker
 
+    # It is not possible to implement __reduce__ for a cdef class. The
+    # two-tuple return cannot handle the keyword-only arguments, and
+    # the three-tuple return (for updating the object's internal state)
+    # does not work either, because cdef classes by default does not have
+    # __dict__. Therefore, the only way to handle keyword-only arguments
+    # for picking a cdef class is to define the following two special
+    # functions, which is in fact preferred over __reduce__.
+
+    def __getstate__(self):
+        cdef dict args = {'code': self.code,
+                          'name': self.name,
+                          'options': self.options,
+                          'backend': self.backend,
+                          'translate_cucomplex': self.translate_cucomplex,
+                          'file_path': self.file_path,
+                          'name_expressions': self.name_expressions}
+        args['enable_cooperative_groups'] = self.enable_cooperative_groups
+        return args
+
+    def __setstate__(self, dict args):
+        self.code = args['code']
+        self.name = self.__name__ = args['name']
+        self.options = args['options']
+        self.backend = args['backend']
+        self.translate_cucomplex = args['translate_cucomplex']
+        self.enable_cooperative_groups = args['enable_cooperative_groups']
+        self.file_path = args['file_path']
+        self.name_expressions = args['name_expressions']
```

Comment:
`args` is generated through the pickling procedure, which is in our full control via `__getstate__`, so all of the keys are always given.

Reply:
Yes currently, but some keys might be added or removed in the future. I'm wondering if the problem will occur when the pickled data of older or newer version is set. Do you think it should raise an error, or ignore the unknown key in such a situation? :thinking:

Reply:
Oh, I see your concern. My current thought is in such events `__getstate__` and `__setstate__` should also be updated to reflect any changes in `RawKernel`'s internal state. But, it means we won't provide strong guarantee that pickles can be reused across major versions. 

I tend to think this is reasonable, because if we add or remove a key, the corresponding compilation procedure must also be changed, and to provide a strong backward compatibility it requires more changes beyond merely touching `RawKernel` as being done here, since unpickling requires a re-compilation.

Reply:
If we don't guarantee the reuse of pickled data across major versions, I think an error should be raised. The compilation procedure might be changed even if the set of the keys is not changed certainly, but we can notify users that the compilation procedure is changed when keys are added or removed at least.

Reply:
How about we add a "raw_ver = 1" to its internal state, compare it when unpickling, and raise an error if unequal or such state is not found? raw_ver will be bumped every time there is a substantial change (though very unlikely).

Reply:
It looks much better!

Reply:
Done in 4c70295. @asi1024 PTAL.